### PR TITLE
Expectation Over Transformation Wrapper

### LINKED ISSUE #191: Potential Feature Request: Expectation over Transformation Attack
One defense that has been discussed before is performing randomized transformations of the input, and [Expectation over Transformations](https://arxiv.org/pdf/1707.07397.pdf) is one counter to this attack. The idea being to simply sample the model for its gradient multiple times, and averaging the gradient estimates to counter the defender using randomized transformations. 

I think this could be implemented with the `ModelWithEstimatedGradients` class in foolbox, but I'm not entirely sure. I was looking to get feedback if this was on the right track - the documentation is a bit light on what exactly the `gradient_estimator` should be returning.

### LINKED ISSUE #563: DifferentiableModelWrapper equivalent for Expectation Over Transformation?
Following what was discussed in #191, I'm trying to implement Expectation Over Transformation in a similar fashion. The original poster mentioned using DifferentiableModelWrapper to do so, but that is obviously no longer available. The only wrapper class that seems to be available currently is a ThresholdingWrapper. I was wondering if anyone could help point me in the right direction for accomplishing this. 

Ideally, I would think implementing EOT would involve having a wrapper around a PyTorchModel object that basically averages the gradients from that model then returns the final, averaged gradients as per EOT.

### PR DESCRIPTION
**Pull request by SerezD**
Closes #563 and #191

My approach for implementing [EoT](https://arxiv.org/pdf/1707.07397.pdf). 
I tried to follow the suggestions of @zimmerrol. 

Basically, I created a new ExpectationOverTransformation Wrapper class that computes the forward pass `n_steps` times (default to 16) and returns the average prediction. 

I also created a new example for pytorch `resnet-18` model, which seems to work. 

**Potential Issue:** since the forward pass is executed `n_steps` times, the attack takes a long time for execution.

### CODE REVIEW

**Comment by SerezD** [comment]
By looking at the other test files, I have added some tests to assert that wrapping the model with EoT does not change clean / robust accuracy (when no random transforms are applied). Don't know if this is correct.

Also, I am not really sure how to add more complex tests (with random transformations applied to the input).

File: tests/test_eot_attack.py
Comment (zimmerrol) [inline tests/test_eot_attack.py:1]: Let's rename this file to something like `test_eot_wrapper.py`.

File: tests/test_eot_attack.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+from foolbox import accuracy
+from foolbox.attacks import LinfBasicIterativeAttack, L1BasicIterativeAttack, L2BasicIterativeAttack
+from foolbox.models import ExpectationOverTransformationWrapper
+from foolbox.types import L2, Linf
+
+from conftest import ModeAndDataAndDescription
+
+
+def test_eot_attack(
```
Comment (zimmerrol) [inline tests/test_eot_attack.py:13]: See comment above.

File: tests/test_eot_attack.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    rand_acc = accuracy(rand_model, x, y)
+    assert acc - rand_acc == 0
+
+    # test with base attacks
+    # (accuracy should not change, since fmodel is not random)
+    attacks = (L1BasicIterativeAttack(), L2BasicIterativeAttack(), LinfBasicIterativeAttack())
+    epsilons = (5000.0, L2(50.), Linf(1.))
+
+    for attack, eps in zip(attacks, epsilons):
+
```
Comment (zimmerrol) [inline tests/test_eot_attack.py:37]: Remove unnecessary whitespace.

**Review by zimmerrol — CHANGES_REQUESTED** [review]
Except a few smaller things, this looks good to me!

**Comment by SerezD** [comment]
I have uploaded a refactored version, which should now be compatible with `black` and `mypy` requirements. This should fix the failed checks for the Tests part.

File: foolbox/models/wrappers.py [hunk 3]
```diff
@@ -21,3 +21,29 @@ def __call__(self, inputs: T) -> T:
         y = ep.where(x < self._threshold, min_, max_).astype(x.dtype)
         z = self._model(y)
         return restore_type(z)
+
+
+class ExpectationOverTransformationWrapper(Model):
```
Comment (zimmerrol) [inline foolbox/models/wrappers.py:26]: Sorry for not mentioning this earlier, but can you please add docstrings to the class & module like its done for the other classes in the project?

**Review by zimmerrol — CHANGES_REQUESTED** [review]
Only one minor comment left, then we can merge this!

**Review by zimmerrol — APPROVED** [review]

**Pull request merged**